src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3581,6 +3581,10 @@ protected BoundExpression BindObjectCreationExpression(ObjectCreationExpressionS
                         diagnostics.Add(ErrorCode.ERR_UnsafeTypeInObjectCreation, node.Location, type));
                     goto case TypeKind.Class;
 
+                case TypeKind.Array:
+                    // ex: new ref[]
+                    return BadExpression(node, LookupResultKind.NotCreatable);
```

Comment:
>return BadExpression(node, LookupResultKind.NotCreatable); [](start = 20, length = 58)

Instead of bailing out like that, we should continue binding for SemanticModel and IOperation benefits. For example, we might go into BindClassCreationExpression (because an array is a class) and handle the array type there specially, if that is even required (we might just get an acceptable error about missing constructor, etc.)

Reply:
See how pointers are handled, the same approach might work here.

---
In reply to: [175167784](https://github.com/dotnet/roslyn/pull/25505#discussion_r175167784) [](ancestors = 175167784)

Reply:
`BindClassCreationExpression` ends up returning a bad expression anyways (line 4683) :-(
I can do the change here, but it's not going to help with semantic model or IOperation.
If we want to eliminate `BadExpression`, it's probably best to do as a separate effort.

Reply:
> BindClassCreationExpression ends up returning a bad expression anyways (line 4683) :-(

The goal is not to avoid producing bad expression, the goal is to not drop constituent pieces and it looks like the bad expression produced at line 4683 doesn't drop them. So they will be available to SemanticModel and IOperation. Tests should confirm that 

---
In reply to: [175974841](https://github.com/dotnet/roslyn/pull/25505#discussion_r175974841) [](ancestors = 175974841)